chain/src/types.rs

```diff
@@ -143,8 +165,8 @@ impl SyncState {
 	}
 
 	/// Get sync error
-	pub fn sync_error(&self) -> Arc<RwLock<Option<Error>>> {
-		Arc::clone(&self.sync_error)
+	pub fn sync_error(&self) -> RwLockReadGuard<Option<Error>> {
+		self.sync_error.read()
 	}
```

Comment:
It would be nice if we could avoid the `RwLockReadGuard` here with something along the lines of what we do with `peers` in p2p -

```
	pub fn sync_error(&self) -> Option<Error> {
		self.sync_error.read().as_ref().cloned()
	}
```

But it looks like the error is not actually cloneable...

Reply:
This is a tricky thing, initially I went for the clone approach, because I agree that returning Guard may look strange, but if we return a detached copy the rwlock guarantees are lost for us, a client gets a snapshot of maybe obsolete state. This is not a big deal in the current code, but in general it doesn't sound great. We have the same situation with state filed, we return a copy, I'm not sure if it's the right approach, was thinking should we remove `Clone` impl or not. We may end up dealing with a wrong state for some period of time.

Reply:
Yeah - we either retain the lock and prevent the case of seeing stale state (or error).
Or we take a "snapshot" of the current state (or error) and return that, and allow the lock to be releases.

I'm not sure we gain much by retaining the lock here.

Reply:
`sync_error` may be harmless, but it all depends on the client behavior, which may change in the future. `status` is more interesting, eg here we have a racy check https://github.com/mimblewimble/grin/blob/master/servers/src/grin/sync/state_sync.rs#L141 `update` may run against a different state. At the same time exposing Guard in `status()` would lead to a deadlock.